src/csharp/Microsoft.Spark.E2ETest/IpcTests/BroadcastTests.cs

```diff
@@ -0,0 +1,167 @@
+                "world first broadcast, 1" };
+
+            Row[] actualRows = _df.Select(testBroadcast(_df["_1"])).Collect().ToArray();
+            string[] actual = actualRows.Select(s => s[0].ToString()).ToArray();
+            Assert.Equal(expected, actual);
+        }
+
+        /// <summary>
+        /// Test Broadcast support by using multiple broadcast variables in a UDF.
+        /// </summary>
+        [Fact]
+        public void TestMultipleBroadcastWithoutEncryption()
+        {
+            var object1ToBroadcast = new TestBroadcastVariable(
+                1,
+                "first broadcast");
+            var object2ToBroadcast = new TestBroadcastVariable(
+                2,
+                "second broadcast");
+
+            Broadcast<TestBroadcastVariable> bc1 = _spark.SparkContext.Broadcast(object1ToBroadcast);
+            Broadcast<TestBroadcastVariable> bc2 = _spark.SparkContext.Broadcast(object2ToBroadcast);
+
+            Func<Column, Column> testBroadcast = Udf<string, string>(
+                str => str +
+                (bc1.Value()).StringValue +
+                " and " +(bc2.Value()).StringValue);
+
+            string[] expected = new[] {
+                "hello first broadcast and second broadcast",
+                "world first broadcast and second broadcast" };
+
+            Row[] actualRows = _df.Select(testBroadcast(_df["_1"])).Collect().ToArray();
+            string[] actual = actualRows.Select(s => s[0].ToString()).ToArray();
+            Assert.Equal(expected, actual);
+
+        }
+
+        /// <summary>
+        /// Test Broadcast.Destroy() that destroys all data and metadata related to the broadcast
+        /// variable and makes it inaccessible from workers.
+        /// </summary>
+        [Fact]
+        public void TestDestroy()
+        {
+            var objectToBroadcast = new TestBroadcastVariable(
+                3,
+                "Broadcast.Destroy()");
+
+            Broadcast<TestBroadcastVariable> bc = _spark.SparkContext.Broadcast(objectToBroadcast);
+
+            Func<Column, Column> testBroadcast = Udf<string, string>(
+                str => str +
+                (bc.Value()).StringValue +
+                ", " + (bc.Value()).IntValue);
+
+            string[] expected = new[] {
+                "hello Broadcast.Destroy(), 3",
+                "world Broadcast.Destroy(), 3" };
+
+            Row[] actualRows = _df.Select(testBroadcast(_df["_1"])).Collect().ToArray();
+            string[] actual = actualRows.Select(s => s[0].ToString()).ToArray();
+            Assert.Equal(expected, actual);
+
+            bc.Destroy();
+
+            try
+            {
+                Row[] rowsAfterDestroy = _df.Select(testBroadcast(_df["_1"])).Collect().ToArray();
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"Broadcast variable accessed after being destroyed: {e.Message}");
+            }
```

Comment:
@imback82 
Catching the exception in this manner because using the following: 
`Assert.Throws<Exception>(() => _df.Select(testBroadcast(_df["_1"])).Collect().ToArray());` -> throws the following error:
`System.Runtime.Serialization.SerializationException: 'Type 'Microsoft.Spark.E2ETest.IpcTests.BroadcastTests' in Assembly 'Microsoft.Spark.E2ETest, Version = 0.9.0.0, Culture = neutral, PublicKeyToken = cc7b13ffcd2ddd51' is not marked as serializable.' `
While trying to serialize the UDF `testBroadcast`. Need some help with debugging this, not sure why `Assert.Throws()` is causing failure in serializing the UDF.

Reply:
Did you try to run it from command line? I don't see this but 
```
20/03/08 19:05:45 ERROR Utils: Exception encountered
org.apache.spark.SparkException: Attempted to use Broadcast(0) after it was destroyed (destroy at NativeMethodAccessorIm                                                                                           pl.java:0)
        at org.apache.spark.broadcast.Broadcast.assertValid(Broadcast.scala:144)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$writeObject$1.apply$mcV$sp(TorrentBroadcast.scala:203)                                                                                                     at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$writeObject$1.apply(TorrentBroadcast.scala:202)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$writeObject$1.apply(TorrentBroadcast.scala:202)
        at org.apache.spark.util.Utils$.tryOrIOException(Utils.scala:1326)
```
You can share the callstack. Btw, I don't think you can reuse the udf since it uses the broadcast variable that is destroyed. I thought we discussed not to call the same udf but to introduce a new udf that doesn't use the broadcast variable to make sure this scenario is fine (this was not working). Something changed?